proposals/p5337.md

```diff
@@ -795,16 +790,57 @@ final match_first {
 fn G(U:! Y) {
   // Can assume `U impls J` and `U.(J.A)` is `bool`.
 }
+
+fn H(V:! W) {
+  // Don't know if `V impls Y`, so we don't know
+  // which `impl` is selected. We use a symbolic
+  // witness for `V` and `V.(J.A)` is unknown
+  // (despite that it would be `bool` in either
+  // case).
+}
 ```
 
-> FIXME _**QUESTION**_: We could say that in this situation the value of `A.J`
-> is known to be `bool` if the query matches `[V:! W] V as J`, even though that
-> `impl` may not be used. The logic is that if that impl is not used, the
-> `impl forall [U:! Y] U as J` will be selected instead and it sets `A.J` to the
-> same thing. This is a bit like what we did in
-> [proposal #2868](https://github.com/carbon-language/carbon-lang/pull/2868)
-> which allowed an `impl` to overlap a `final impl` as long as agreed on the
-> overlap.
+The compiler will only use a known matching final `impl`'s associated constants
+if it is the first that _could_ match, as determined by type structure. For
+example:
+
+```carbon
+interface L(T:! type) {
+  let B:! type;
+}
+class C(T:! type) {}
+
+final match_first {
+  // Can use the associated constants from this impl,
+  // since it is the first, so nothing could match
+  // ahead of it.
+  impl forall [U:! Y] C(U) as L(bool) where .B = i32;
+
+  // Can use the associated constants from this impl
+  // only if V can't be `C(U)`, and so it is known
+  // that it won't use the previous impl instead.
+  impl forall [V:! W] V as L(bool) where .B = bool;
+
+  // Can use the associated constants from this impl,
+  // since a query for `T as L(i32)` can't match any
+  // of the earlier impls in the `final match_first`.
+  impl forall [T:! Z] T as L(i32) where .B = f32;
+}
+
+class D {}
+impl D as W {}
+// D as L(bool) definitely uses the middle `impl forall`
+// from the `final match_first` so `(D as L(bool)).B` is
+// `bool`.
+let x: (D as L(bool)).B = true;
+
+class E(T:! type) {}
+impl forall [T:! type] C(E(T)) as W {}
+// `C(E(T))` does impl `W`, so it impls `L(bool)`. But
+// since we don't know whether `E(T)` impls `Y`, we
+// don't know which impl it will use, and so
+// `(C(E(T)) as L(bool)).B` is unknown.
```

Comment:
As above, this means the result will not be a concrete ImplWitness here, it would have to be a LookupImplWitness even though it landed at a final impl?

Will we be able to diagnose this meaningfully when the user tries to use a concrete thing out of a `final match_first` and we deny them?

Why do we go with a non-concrete result instead of relying on poisoning to prevent the value from changing? Would that not work here due to cross-file concerns?

Reply:
This is indded an example of no concrete ImplWitness. However, I do not think we should be doing poisoning. Have I left something in the proposal to say that there should be some poisoning involved in this situation?

Reply:
No, but it's not discussed as an alternative. It would allow use of a final impl to get a concrete result, so that seems like something that could be discussed then, as I find it surprising to select a final impl but not get a concrete ImplWitness back.

We have already said we should poison concrete results from generic queries elsewhere: https://github.com/carbon-language/carbon-lang/blob/8c4ff33cb16873effd09768a6916a9d373af0bc7/toolchain/check/testdata/impl/lookup/min_prelude/specialization_poison.carbon#L145-L171